sdk/metrics/src/test/java/io/opentelemetry/sdk/metrics/CardinalityTest.java

```diff
@@ -118,7 +118,7 @@ void staleMetricsDropped_synchronousInstrument() {
                                     (Consumer<SumData<LongPointData>>)
                                         sumPointData ->
                                             assertThat(sumPointData.getPoints().size())
-                                                .isEqualTo(10))));
+                                                .isEqualTo(2000))));
```

Comment:
not clear to me...is this a significant behavioral change due to this refactoring?

Reply:
What I mean by that is... is this change in behavior going to be a breaking change for anyone? Seems like suddenly getting 20x the number of points could be significant to some backends.

Reply:
The behavior has changed, but the current behavior is odd and the new behavior is slightly less odd so I think its defensible. 

Currently, if the temporality is cumulative, we:
- Record data in delta temporality, [limiting the storage](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/DefaultSynchronousMetricStorage.java#L110-L119) to 2000 distinct attributes. Any attributes in excess of 2000 are dropped.
- Merge the [deltas into the cumulatives](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/TemporalMetricStorage.java#L79)
- Limit the merged cumulative to 2000 distinct attributes. If after merging the delta into the cumulatives, we exceed the 2000 limit, trim the cumulatives down to only those [seen since the last collection](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/TemporalMetricStorage.java#L82-L84). 

So there's two rounds of enforcement of the cardinality limit and the behavior isn't very intuitive / expected. 

The new behavior is much simpler: limit to 2000 attributes in the storage. Drop all measurements which exceed that. This makes the behavior of cardinality enforcement the same with cumulative and delta. I don't think this will be surprising / breaking for any users, since the change is limited to reducing cardinality after the limit has already been hit. I.e. you'll see a potential reduction in cardinality only after you've hit cardinality limits. 

There's some work [active in the spec](https://github.com/open-telemetry/opentelemetry-specification/pull/2960) which would change this behavior further. The most likely outcome is If a configurable cardinality limit is exceeded, log a warning and emit a single point with a static attribute key / value which aggregates measurements across all dimensions.